openmls/src/framing/message.rs

```diff
@@ -1,127 +1,212 @@
+use std::io::BufReader;
+
+use tls_codec::{Deserialize, Serialize};
+
 use super::*;
 
-/// Unified message type for incoming MLS messages.
+/// Unified message type for MLS messages.
 /// Since the memory footprint can differ considerably between [`VerifiableMlsPlaintext`]
 /// and [`MlsCiphertext`], we use `Box<T>` for more efficient memory allocation.
-#[derive(Debug, Clone)]
-pub enum MlsMessageIn {
-    /// An OpenMLS `VerifiableMlsPlaintext`.
+/// This is only used internally, externally we use either [`MlsMessageIn`] or
+/// [`MlsMessageOut`], depending on the context.
+#[derive(PartialEq, Debug, Clone)]
+pub(crate) enum MlsMessage {
+    /// Plaintext message
     Plaintext(Box<VerifiableMlsPlaintext>),
 
-    /// An OpenMLS `MlsCiphertext`.
+    /// Ciphertext message
     Ciphertext(Box<MlsCiphertext>),
 }
 
-impl MlsMessageIn {
+impl MlsMessage {
     /// Get the wire format
-    pub fn wire_format(&self) -> WireFormat {
+    fn wire_format(&self) -> WireFormat {
         match self {
-            MlsMessageIn::Ciphertext(_) => WireFormat::MlsCiphertext,
-            MlsMessageIn::Plaintext(_) => WireFormat::MlsPlaintext,
+            MlsMessage::Ciphertext(_) => WireFormat::MlsCiphertext,
+            MlsMessage::Plaintext(_) => WireFormat::MlsPlaintext,
         }
     }
 
     /// Get the group ID
-    pub fn group_id(&self) -> &GroupId {
+    fn group_id(&self) -> &GroupId {
         match self {
-            MlsMessageIn::Ciphertext(m) => m.group_id(),
-            MlsMessageIn::Plaintext(m) => m.group_id(),
+            MlsMessage::Ciphertext(m) => m.group_id(),
+            MlsMessage::Plaintext(m) => m.group_id(),
         }
     }
 
     /// Get the epoch
-    pub fn epoch(&self) -> GroupEpoch {
+    fn epoch(&self) -> GroupEpoch {
         match self {
-            MlsMessageIn::Ciphertext(m) => m.epoch(),
-            MlsMessageIn::Plaintext(m) => m.epoch(),
+            MlsMessage::Ciphertext(m) => m.epoch(),
+            MlsMessage::Plaintext(m) => m.epoch(),
         }
     }
 
     /// Get the content type
-    pub fn content_type(&self) -> ContentType {
+    fn content_type(&self) -> ContentType {
         match self {
-            MlsMessageIn::Ciphertext(m) => m.content_type(),
-            MlsMessageIn::Plaintext(m) => m.content_type(),
+            MlsMessage::Ciphertext(m) => m.content_type(),
+            MlsMessage::Plaintext(m) => m.content_type(),
         }
     }
 
     /// Returns `true` if this is a handshake message and `false` otherwise.
-    pub fn is_handshake_message(&self) -> bool {
+    fn is_handshake_message(&self) -> bool {
         self.content_type().is_handshake_message()
     }
+
+    /// Tries to deserialize from a byte slice. Returns [`MlsMessageError::DecodingError`] on failure.
+    fn try_from_bytes(bytes: &[u8]) -> Result<Self, MlsMessageError> {
+        let mut reader = BufReader::new(bytes);
```

Comment:
Is using `BufReader` more efficient than just using the slice?
If not, why not just call `tls_deserialize` on the slice?

Reply:
The problem with using it on the slice directly is that the slice needs to be mutable and I don't think that's a great requirement from a consumer perspective. I have replaced this with a `Cursor` now, because I realised that a `BufReader` would make a copy of the data.

Reply:
The slice doesn't need to be mutable. I also get confused by this sometimes. But the slice is only a pointer into a structure that's fine to modify. So something like this should work.
```rust
    fn try_from_bytes(mut bytes: &[u8]) -> Result<Self, MlsMessageError> {
        MlsMessage::tls_deserialize(&mut bytes).map_err(|_| MlsMessageError::DecodingError)
    }
```

Reply:
Nice one!